pkg/apply/taskrunner/task.go

```diff
@@ -137,21 +131,29 @@ func (w *WaitTask) setTimer(taskContext *TaskContext) {
 	}
 }
 
-func amendTimeoutError(taskContext *TaskContext, err error) {
-	if timeoutErr, ok := err.(*TimeoutError); ok {
-		var timedOutResources []TimedOutResource
-		for _, id := range timeoutErr.Identifiers {
-			result := taskContext.ResourceCache().Get(id)
-			if timeoutErr.Condition.Meets(result.Status) {
-				continue
-			}
-			timedOutResources = append(timedOutResources, TimedOutResource{
-				Identifier: id,
-				Status:     result.Status,
-				Message:    result.StatusMessage,
-			})
+// timeoutError returns an error with the objects that have not yet reached the
+// desired condition.
+func (w *WaitTask) timeoutError(taskContext *TaskContext) *TimeoutError {
+	expected := w.pending(taskContext)
+	ids := object.ObjMetadataSet{}
+	tors := []TimedOutResource{}
+	for _, id := range expected {
+		result := taskContext.ResourceCache().Get(id)
```

Comment:
So will this call always return the same version of the resource as was used by the `WaitTask` to determine that it failed to reconcile? I'm wondering if there is a chance where we determine the resource has failed to reconcile before the timeout, but does so just afterwards and ends up not being included here?

Reply:
This timeout error includes the IDs of all the objects that are still not reconciled at timeout.

If it reconciles after timeout, it won't be included. Tho there's a lag between timeout and all events being sent, because the caller needs to be draining the event channel.

If that's something we want to defend against, we could use two for loops, one to assess and make a list of IDs, and another to send the events.

Either way, the TaskResult wont be sent until after this returns. So if it reconciled while sending events it will still count as a timeout error.